app/scripts/lib/indexed-db-backend.ts

```diff
@@ -0,0 +1,185 @@
+            keyPath: ['name', 'chainId'],
+          });
+        }
+      };
+
+      request.onsuccess = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result;
+        const transaction = db.transaction([this.storeName], mode);
+        const objectStore = transaction.objectStore(this.storeName);
+        resolve(objectStore);
+      };
+    });
+  }
+
+  async read(key: StorageKey, checksum: string): Promise<ArrayBuffer> {
+    return new Promise<ArrayBuffer>((resolve, reject) => {
+      this._getObjectStore('readonly')
+        .then((objectStore) => {
+          const request = objectStore.get([key.name, key.chainId]);
+
+          request.onsuccess = async (event) => {
+            const data = (event.target as any)?.result?.data;
+            await validateChecksum(key, data, checksum).catch((error) => {
+              reject(error);
+            });
+            resolve(data);
+          };
+
+          request.onerror = (event) => {
+            reject(
+              new Error(`Error reading data: ${(event.target as any)?.error}`),
+            );
+          };
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  async write(
+    key: StorageKey,
+    data: ArrayBuffer,
+    checksum: string,
+  ): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      validateChecksum(key, data, checksum)
+        .then(() => {
+          this._getObjectStore('readwrite')
+            .then((objectStore) => {
+              const request = objectStore.put({ ...key, data });
+
+              request.onsuccess = () => {
+                resolve();
+              };
+
+              request.onerror = (event) => {
+                reject(
+                  new Error(
+                    `Error writing data: ${(event.target as any)?.error}`,
+                  ),
+                );
+              };
+            })
+            .catch((error) => {
+              reject(error);
+            });
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  public async delete(key: StorageKey): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      this._getObjectStore('readwrite')
+        .then((objectStore) => {
+          const request = objectStore.delete([key.name, key.chainId]);
+
+          request.onsuccess = () => {
+            resolve();
+          };
+
+          request.onerror = (event) => {
+            reject(
+              new Error(`Error deleting data: ${(event.target as any)?.error}`),
+            );
+          };
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  public async dir(): Promise<StorageKey[]> {
+    return new Promise<StorageKey[]>((resolve, reject) => {
+      this._getObjectStore('readwrite')
+        .then((objectStore) => {
+          const request = objectStore.getAllKeys();
+
+          request.onsuccess = (event) => {
```

Comment:
It seems all these methods create promises using the `onsuccess` and `onerror` methods.

Would it simplify this logic and avoid a lot of duplication if we create a private helper method such as `promisifyRequest` which converts any `request` to a `Promise` and then we can convert the majority of these methods to linear `async` code using `awaits`? We could resolve to the `event` or the `event.target.result` if each operation has one.

This method for example could become:

```
const objectStore = await this._getObjectStore('readwrite');
const result = await this._promisifyRequest(objectStore.getAllKeys());
return result.map( ([name, chainId]: string[]) => ({ name, chainId }) );
```

Reply:
I updated the PR, please check.